### PR TITLE
refactoring of returnn jobs

### PR DESCRIPTION
**Pull request by JackTemaki**
- mainly docstrings and adding "out_" prefix
- create a util.create_executable instead of rnn.sh creation

I did not touch training and search yet, as this will come in a separate PR with more critical code changes.

### CODE REVIEW

File: returnn/oggzip.py [hunk 1]
```diff
@@ -33,8 +38,8 @@ def __init__(
         :param int raw_sample_rate: raw audio sampling rate
         :param int feat_sample_rate: feature sampling rate
         :param bool no_conversion: do not call the actual conversion, assume the audio files are already correct
-        :param str|Path returnn_python_exe: path to python binary
-        :param str|Path returnn_root: path to root dir of RETURNN
+        :param Path|str returnn_python_exe: file path to the executable for running returnn (python binary or .sh)
+        :param path|str returnn_root: file path to the RETURNN repository root folder
```
Comment (albertz) [inline returnn/oggzip.py:42]:
```suggestion
        :param Path|str returnn_root: file path to the RETURNN repository root folder
```

File: returnn/compile.py [hunk 2]
```diff
@@ -24,6 +29,18 @@ def __init__(
         returnn_python_exe=None,
         returnn_root=None,
     ):
+        """
+
+        :param Path returnn_config: Path to a RETURNN config file
```
Comment (christophmluscher) [inline returnn/compile.py:34]: maybe rather Path|str ?

File: returnn/config.py [hunk 3]
```diff
@@ -61,6 +76,17 @@ def __init__(
         extra_python_code="",
         extra_python_hash=None
     ):
+        """
+
+        :param dict config:
+        :param dict post_config:
+        :param None|str|Callable|Class|tuple|list|dict python_prolog: str or structure containing str/callables/classes
+            that should be pasted as code at the beginning of the config file
+        :param str python_prolog_hash: a
```
Comment (christophmluscher) [inline returnn/config.py:85]: only a?

**Review by christophmluscher — APPROVED** [review]

**Comment by christophmluscher** [comment]
There was a conflict because I renamed the Job ExtractPriorFromHDF5 to ExtractPriorFromHDF5Job

@JackTemaki please check if still ok.

File: returnn/compile.py (hunk 2 above)
Comment (JackTemaki) [reply returnn/compile.py:34]: It is wrong in general, it should be ReturnnConfig|Path|str, the job accepts also a config object, not only a file.

File: returnn/config.py [hunk 4]
```diff
@@ -61,6 +76,17 @@ def __init__(
         extra_python_code="",
```
Comment (curufinwe) [inline returnn/config.py:76]: This should also be refactored to python_epilog / python_epilog_hash

File: returnn/hdf.py [hunk 5]
```diff
@@ -99,6 +103,11 @@ def hash(cls, parsed_args):
 
 
 class ReturnnRasrDumpHDFJob(ReturnnDumpHDFJob):
+    """
+    This Job extends the ReturnnDumpHDFJob to allow to use ExternSprintDataset.
```
Comment (curufinwe) [inline returnn/hdf.py:107]: * to allow the use of ExternSprintDataset

File: returnn/hdf.py [hunk 6]
```diff
@@ -108,6 +117,15 @@ def __init__(
         buffer_size=200 * 1024,
         **kwargs,
     ):
+        """
+
+        :param crp: common RASR parameters
```
Comment (curufinwe) [inline returnn/hdf.py:122]: type missing should be rasr.CommonRasrParameters

File: util.py [hunk 7]
```diff
@@ -229,3 +230,26 @@ def remove_unwanted_whitespace(elem):
 
     with uopen(filename, "w") as f:
         f.write(xml_string)
+
+
+def create_executable(filename, command):
```
Comment (curufinwe) [inline util.py:235]: not used in this commit?

File: adaptation/linear_adaptation_layer.py [hunk 8]
```diff
@@ -232,10 +232,10 @@ def setup_training(self):
                 **self.returnn_training_args
             )
             self.jobs["train_{}".format(key)] = j
-            self.models[key] = j.models
-            self.add_input(j.plot_lr)
+            self.models[key] = j.out_models
```
Comment (curufinwe) [inline adaptation/linear_adaptation_layer.py:235]: not refactored in this commit?

File: meta/system.py [hunk 9]
```diff
@@ -610,8 +610,8 @@ def train_nn(
             **kwargs
         )
         self.jobs[feature_corpus]["train_nn_%s" % name] = j
-        self.nn_models[feature_corpus][name] = j.models
-        self.nn_configs[feature_corpus][name] = j.returnn_config_file
+        self.nn_models[feature_corpus][name] = j.out_models
```
Comment (curufinwe) [inline meta/system.py:613]: not refactored in this commit?

**Review by curufinwe — CHANGES_REQUESTED** [review]

**Comment by JackTemaki** [comment]
I excluded the training because there are more critical changes, I now checked that the rest is not edited wrongly in this PR.

**Review by curufinwe — CHANGES_REQUESTED** [review]

File: returnn/config.py [hunk 10]
```diff
@@ -58,18 +73,29 @@ def __init__(
...
+        :param str python_epilog_hash: sets a specific hash for the python_epilog
+        """
         self.config = config
         self.post_config = post_config if post_config is not None else {}
         self.python_prolog = python_prolog
         self.python_prolog_hash = (
             python_prolog_hash if python_prolog_hash is not None else python_prolog
         )
-        self.extra_python_code = extra_python_code
+        self.extra_python_code = python_epilog
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply returnn/config.py]: Ah yes, sorry

File: returnn/config.py [hunk 11]
```diff
@@ -105,7 +131,7 @@ def serialize(self):
             config_lines.append("config = {}")
 
         prolog = self.__parse_python(self.python_prolog)
-        extra_python_code = self.__parse_python(self.extra_python_code)
+        extra_python_code = self.__parse_python(self.python_epilog)
```
Comment (curufinwe) [inline returnn/config.py:134]: here too

File: returnn/config.py [hunk 12]
```diff
@@ -160,25 +186,33 @@ def __parse_python(self, code, name=None):
         raise RuntimeError("Could not serialize %s" % code)
 
     def hash(self):
-        h = {"returnn_config": self.config, "extra_python_hash": self.extra_python_hash}
+        h = {"returnn_config": self.config, "extra_python_hash": self.python_epilog_hash}
```
Comment (curufinwe) [inline returnn/config.py:189]: here too

**Review by curufinwe — CHANGES_REQUESTED** [review]
What I meant was to change all occurrences of extra_python. It should now only be known as python_epilog.

File: returnn/config.py (hunk 12 above)
Comment (JackTemaki) [reply returnn/config.py]: I should get a coffee... yes obviously this needs to be changed.

**Review by curufinwe — APPROVED** [review]

**Pull request merged**